compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -292,7 +292,12 @@ fn impl_intersection_has_impossible_obligation<'cx, 'tcx>(
             Obligation::new(infcx.tcx, ObligationCause::dummy(), param_env, predicate)
         })
         .chain(obligations)
-        .find(|o| !selcx.predicate_may_hold_fatal(o));
+        .find(|o| {
+            selcx.evaluate_root_obligation(o).map_or(
+                true, // Overflow has occurred, and treat the obligation as possibly holding.
```

Comment:
An overflow is *not* an failing obligation. It is an ambiguity, and this is coherence, which must *never* treat an overflow as an error for soundness reasons.

Reply:
Understood. Thank you for explaining that.

Reply:
Specifically, an overflow represents a state of "we don't know whether or not a predicate holds because we didn't iterate deep enough" (where "deep enough" is sometimes infinitely deep -- as in this case -- or sometimes just because we set the recursion depth too low).

In coherence, we must never treat "we don't know" as "certainly does not hold", because in this code, "certainly does not hold" is used as proof that two impls may not overlap.

Elsewhere in the compiler, outside of coherence, we can use overflows as "probably does not hold" because even if it did hold, we'd never be able to instantiate it during codegen. That's why other code, such as `evaluate_obligation_no_overflow` treats this case as an error. The worst it can do is affect inference or the choice of a method during method dispatch.

Reply:
Thank you so much for your clear and concise explanation. I now have a better understanding of this topic.